README.pod
=pod

=encoding UTF-8

=for :stopwords Ferenc Erki adjust GmbH CMDB CPAN Flink hadoop Hadoop libvirt scala VM YAML

=head1 NAME

Rex::Flink - automate Flink tasks with Rex

=head1 VERSION

version v0.5.0

=head1 SYNOPSIS

    # Rexfile
    use Rex::Flink;

    # CLI
    $ rex Flink:setup
    $ rex Flink:start
    $ rex Flink:stop

=head1 DESCRIPTION

This module automates Flink tasks with Rex.

B<Please note this is a work still heavily in progress, the interface might be subject to change before version 1.0.0.>

=head1 TASKS

=head2 Flink:setup

Sets Flink up from binary tarball. The installation directory can be controlled from CMDB via C<install_path>. Default is C<~/flink>.

Supported options:

=over 4

=item --hadoop

Hadoop version (default: C<27>)

=item --scala

Scala version (default: C<2.11>)

=item --version

Flink version (default: C<1.2.0>)

=back

=head2 Flink:configure

Configures Flink from a file, which is treated as a standard Rex template. The name of the file to be deployed can be controlled from CMDB via C<config_file>. Default is C<templates/flink-conf.yaml>.

=head2 Flink:start

Starts a Flink service.

Supported options:

=over 4

=item --service

Service name to start (default: C<jobmanager>)

=item --mode

Execution mode to start the service with (default: C<local>)

=back

=head2 Flink:stop

Stops a Flink service.

Supported options:

=over 4

=item --service

Service name to stop (default: C<jobmanager>)

=back

=head2 Flink:Pipeline:upload

Uploads a compiled pipeline to the remote machine from the local one. By default it uploads into C<~/flink/current/pipelines>.

Supported options:

=over 4

=item --file

Local filename of the pipeline to upload.

=back

=head2 Flink:Pipeline:run

Runs a pipeline on the remote machine.

Supported options:

=over 4

=item --file

Remote filename of the pipeline to run. Mandatory, without any default value.

=item --class

Class with the pipeline entry point.

=item --parallelism

Parallelism for the pipeline.

=item --options

Any extra options to be passed to the pipeline.

=back

=head1 DIAGNOSTICS

=over 4

=item *

C<No pipeline file specified>

=back

No pipeline file is specified to run with C<Flink:Pipeline:run> task.

=head1 CONFIGURATION AND ENVIRONMENT

This module can use configuration specified under the C<flink> node in the CMDB. Example YAML structure:

 flink:
     config_file: templates/flink-conf.yaml
     hadoop: 27
     install_path: ~/flink
     scala: 2.11
     version: 1.2.0

This module does not use any environment variables. See also L<TESTING>.

=head1 DEPENDENCIES

This module depends on the modules detailed below.

CPAN modules:

=over 4

=item *

Const::Fast

=item *

Rex

=back

=head1 INCOMPATIBILITIES

There are no known incompatibilities with other modules.

=head1 BUGS AND LIMITATIONS

There are no known bugs. Make sure they are reported.

=head1 TESTING

This module is capable to do testing with the help of the C<Rex::Test> module inside a VM via libvirt.

A URL pointing to a VM image to be used for the testing should be specified via the C<REX_FLINK_TEST_IMAGE> environment variable. For example:

 # CLI
 $ REX_FLINK_TEST_IMAGE='file:///var/lib/livirt/images/image.qcow2' rex Test:run
 $ REX_FLINK_TEST_IMAGE='https://domain.tld/image.qcow2' rex Test:run

=head1 AUTHOR

Ferenc Erki <ferki@adjust.com>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2017 by adjust GmbH.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Changes
Revision history for Rex-Flink

{{$NEXT}}
 [API CHANGES]

 [BUG FIXES]

 [DOCUMENTATION]

 [ENHANCEMENTS]

 [MAJOR]

 [MINOR]

 [NEW FEATURES]

 [REVISION]

v0.5.0 2017-04-12T16:04:21Z
 [BUG FIXES]
 - Update download URL

 [ENHANCEMENTS]
 - Verify MD5 checksum of downloaded tarball

v0.4.0 2017-04-12T12:51:31Z
 [ENHANCEMENTS]
 - Support individual Flink services

v0.3.0 2017-04-11T12:58:39Z
 [BUG FIXES]
 - Increase RAM for test VM

 [DOCUMENTATION]
 - Warn about WIP status
 - General fixes

 [MINOR]
 - Move common code into a separate module

 [NEW FEATURES]
 - Initial support for pipeline management

v0.2.0 2017-04-07T15:27:58Z
 [BUG FIXES]
 - Use share directory for sample CMDB

 [DOCUMENTATION]
 - Document install_path

 [ENHANCEMENTS]
 - Add basic configure task

v0.1.0 2017-04-06T14:05:07Z
 - Initial release


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;



use ExtUtils::MakeMaker;

use File::ShareDir::Install;
$File::ShareDir::Install::INCLUDE_DOTFILES = 1;
$File::ShareDir::Install::INCLUDE_DOTDIRS = 1;
install_share dist => "share";


my %WriteMakefileArgs = (
  "ABSTRACT" => "automate Flink tasks with Rex",
  "AUTHOR" => "Ferenc Erki <ferki\@adjust.com>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::ShareDir::Install" => "0.06"
  },
  "DISTNAME" => "Rex-Flink",
  "LICENSE" => "perl",
  "NAME" => "Rex::Flink",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Const::Fast" => 0,
    "File::ShareDir" => 0,
    "Rex" => 0,
    "Rex::CMDB" => 0,
    "Rex::Commands::MD5" => 0,
    "Rex::Commands::Upload" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "v0.5.0",
  "test" => {
    "TESTS" => ""
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Const::Fast" => 0,
  "File::ShareDir" => 0,
  "Rex" => 0,
  "Rex::CMDB" => 0,
  "Rex::Commands::MD5" => 0,
  "Rex::Commands::Upload" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);

{
package
MY;
use File::ShareDir::Install qw(postamble);
}


lib/Rex/Flink/Common.pm
package Rex::Flink::Common;

# ABSTRACT: common code for Rex::Flink automation

use strict;
use warnings;

our $VERSION = '9999';

use File::ShareDir 'dist_dir';
use Rex -base;
use Rex::CMDB;

my $dist_share_dir = dist_dir('Rex-Flink');

if ( !-d 'cmdb' ) {    # if there is no local CMDB
    my $cmdb_path;
    if ( -d 'share/cmdb' ) {
        $cmdb_path = 'share/cmdb';    # use the CMDB under share if exists
    }
    else {
        # use the installed CMDB otherwise
        $cmdb_path = "$dist_share_dir/cmdb";
    }

    set cmdb => { type => 'YAML', path => $cmdb_path };
}

set path_map => { 'templates/' =>
      [ 'templates/', "$dist_share_dir/templates", 'share/templates' ], };

my $flink = get cmdb 'flink';

my $install_path = $flink->{install_path} // '~/flink';
my $flink_dir    = "$install_path/current";
my $flink_bin    = "$flink_dir/bin/flink";
my $pipeline_dir = "$flink_dir/pipelines";

set flink => {
    cmdb         => $flink,
    install_path => $install_path,
    flink_dir    => $flink_dir,
    flink_bin    => $flink_bin,
    pipeline_dir => $pipeline_dir,
};

1;

__END__

=for :stopwords CMDB namespace

=head1 SYNOPSIS

    use Rex::Flink::Common;

    my $flink = get 'flink';

=head1 DESCRIPTION

This module contains common code for the various modules in the Rex::Flink namespace.

It sets CMDB path and C<path_map>, and also registers the C<flink> configuration parameter.

=head1 DIAGNOSTICS

This module does not do any error checking (yet).

=head1 CONFIGURATION AND ENVIRONMENT

See L<Rex::Flink/CONFIGURATION AND ENVIRONMENT>.

=head1 DEPENDENCIES

See L<Rex::Flink/DEPENDENCIES>.

=head1 INCOMPATIBILITIES

There are no known incompatibilities with other modules.

=head1 BUGS AND LIMITATIONS

There are no known bugs. Make sure they are reported.


lib/Rex/Flink/Pipeline.pm
package Rex::Flink::Pipeline;

# ABSTRACT: pipeline management tasks for Rex::Flink

use strict;
use warnings;

our $VERSION = '9999';

use Carp;
use Rex -base;
use Rex::Commands::Upload;

use Rex::Flink::Common;

my $flink = get 'flink';

desc 'Upload a Flink pipeline';
task 'upload' => sub {
    my $params       = shift;
    my $file         = $params->{file};
    my $pipeline_dir = $flink->{pipeline_dir};

    file $pipeline_dir, ensure => 'directory';

    upload $file, $pipeline_dir;
};

desc 'Run a pipeline';
task 'run' => sub {
    my $params      = shift;
    my $file        = $params->{file};
    my $class       = $params->{class};
    my $parallelism = $params->{parallelism};
    my $job_options = $params->{options} // q{};

    my $pipeline_dir = $flink->{pipeline_dir};
    my $flink_bin    = $flink->{flink_bin};

    croak 'No pipeline file is specified' if !defined $file;

    # compile command
    my @command_parts = qq($flink_bin run);

    if ( defined $class ) {
        push @command_parts, qq(--class $class);
    }

    if ( defined $parallelism ) {
        push @command_parts, qq(--parallelism $parallelism);
    }

    push @command_parts, qq($file $job_options);
    my $command = join q{ }, @command_parts;

    run $command;
};

1;

__END__

=for :stopwords Flink

=head1 SYNOPSIS

    # Rexfile
    use Rex::Flink;

    # CLI
    $ rex Flink:Pipeline:upload --file=pipeline.jar
    $ rex Flink:Pipeline:run --file=remote/path/to/pipeline.jar

=head1 DESCRIPTION

This module contains tasks related to Flink pipeline management.

See L<Rex::Flink/TASKS> for more details.

=head1 DIAGNOSTICS

See L<Rex::Flink/DIAGNOSTICS>.

=head1 CONFIGURATION AND ENVIRONMENT

See L<Rex::Flink/CONFIGURATION AND ENVIRONMENT>.

=head1 DEPENDENCIES

See L<Rex::Flink/DEPENDENCIES>.

=head1 INCOMPATIBILITIES

There are no known incompatibilities with other modules.

=head1 BUGS AND LIMITATIONS

There are no known bugs. Make sure they are reported.


lib/Rex/Flink.pm
package Rex::Flink;

# ABSTRACT: automate Flink tasks with Rex

use strict;
use warnings;

our $VERSION = '9999';

use Carp;
use Const::Fast;
use Rex -base;
use Rex::CMDB;
use Rex::Commands::MD5;
use Rex::Flink::Common;
use Rex::Flink::Pipeline;

my $flink = get 'flink';
const my $SPACE => q{ };

desc 'Setup Flink from tarball';
task 'setup' => sub {
    my $params = shift;

    my $version = $params->{version} // $flink->{cmdb}->{version} // '1.2.0';
    my $hadoop  = $params->{hadoop}  // $flink->{cmdb}->{hadoop}  // '27';
    my $scala   = $params->{scala}   // $flink->{cmdb}->{scala}   // '2.11';

    my $filename = "flink-${version}-bin-hadoop${hadoop}-scala_${scala}.tgz";
    my $url
      = "http://archive.apache.org/dist/flink/flink-${version}/$filename";

    my $tmp_dir  = Rex::Config::get_tmp_dir();
    my $tmp_file = "$tmp_dir/$filename";
    my $md5_file = "$tmp_file.md5";

    run "curl --location $url --output $tmp_file",     creates => $tmp_file;
    run "curl --location $url.md5 --output $md5_file", creates => $md5_file;

    my $md5 = md5($tmp_file);
    my ($expected) = ( split $SPACE, cat $md5_file );
    my $verified = $md5 eq $expected;

    if ( !$verified ) {
        mv $tmp_file, "$tmp_file.bad";
        mv $md5_file, "$md5_file.bad";
        croak "MD5 checksum mismatch for $tmp_file";
    }

    extract $tmp_file, to => $flink->{install_path};

    symlink "$flink->{install_path}/flink-${version}", $flink->{flink_dir};

    needs 'configure';
};

desc 'Configure Flink from a template file';
task 'configure' => sub {
    file "$flink->{flink_dir}/conf/flink-conf.yaml",
      content => template( $flink->{cmdb}->{config_file} );
};

desc 'Start Flink in local mode';
task 'start' => sub {
    my $params  = shift;
    my $service = $params->{service} // 'jobmanager';
    my $mode    = $params->{mode} // 'local';

    my $command = "bin/$service.sh start";
    if ( $service eq 'jobmanager' ) {
        $command .= " $mode";
    }

    run $command, cwd => $flink->{flink_dir};
};

desc 'Stop Flink in local mode';
task 'stop' => sub {
    my $params = shift;
    my $service = $params->{service} // 'jobmanager';

    my $command = "bin/$service.sh stop";

    run $command, cwd => $flink->{flink_dir};
    sleep 1;
};

1;

__END__

=for :stopwords CMDB CPAN Flink hadoop Hadoop libvirt scala VM YAML

=head1 SYNOPSIS

    # Rexfile
    use Rex::Flink;

    # CLI
    $ rex Flink:setup
    $ rex Flink:start
    $ rex Flink:stop

=head1 DESCRIPTION

This module automates Flink tasks with Rex.

B<Please note this is a work still heavily in progress, the interface might be subject to change before version 1.0.0.>

=task Flink:setup

Sets Flink up from binary tarball. The installation directory can be controlled from CMDB via C<install_path>. Default is C<~/flink>.

Supported options:

=begin :list

= --hadoop
Hadoop version (default: C<27>)

= --scala
Scala version (default: C<2.11>)

= --version
Flink version (default: C<1.2.0>)

=end :list

=task Flink:configure

Configures Flink from a file, which is treated as a standard Rex template. The name of the file to be deployed can be controlled from CMDB via C<config_file>. Default is C<templates/flink-conf.yaml>.

=task Flink:start

Starts a Flink service.

Supported options:

=begin :list

= --service
Service name to start (default: C<jobmanager>)

= --mode
Execution mode to start the service with (default: C<local>)

=end :list

=task Flink:stop

Stops a Flink service.

Supported options:

=begin :list

= --service
Service name to stop (default: C<jobmanager>)

=end :list

=task Flink:Pipeline:upload

Uploads a compiled pipeline to the remote machine from the local one. By default it uploads into C<~/flink/current/pipelines>.

Supported options:

=begin :list

= --file
Local filename of the pipeline to upload.

=end :list

=task Flink:Pipeline:run

Runs a pipeline on the remote machine.

Supported options:

=begin :list

= --file
Remote filename of the pipeline to run. Mandatory, without any default value.

= --class
Class with the pipeline entry point.

= --parallelism
Parallelism for the pipeline.

= --options
Any extra options to be passed to the pipeline.

=end :list

=head1 DIAGNOSTICS

=for :list
* C<No pipeline file specified>

No pipeline file is specified to run with C<Flink:Pipeline:run> task.

=head1 CONFIGURATION AND ENVIRONMENT

This module can use configuration specified under the C<flink> node in the CMDB. Example YAML structure:

 flink:
     config_file: templates/flink-conf.yaml
     hadoop: 27
     install_path: ~/flink
     scala: 2.11
     version: 1.2.0

This module does not use any environment variables. See also L<TESTING>.

=head1 DEPENDENCIES

This module depends on the modules detailed below.

CPAN modules:

=for :list
* Const::Fast
* Rex

=head1 INCOMPATIBILITIES

There are no known incompatibilities with other modules.

=head1 BUGS AND LIMITATIONS

There are no known bugs. Make sure they are reported.

=head1 TESTING

This module is capable to do testing with the help of the C<Rex::Test> module inside a VM via libvirt.

A URL pointing to a VM image to be used for the testing should be specified via the C<REX_FLINK_TEST_IMAGE> environment variable. For example:

 # CLI
 $ REX_FLINK_TEST_IMAGE='file:///var/lib/livirt/images/image.qcow2' rex Test:run
 $ REX_FLINK_TEST_IMAGE='https://domain.tld/image.qcow2' rex Test:run
